Ore Samples can require tools; OrePart bugfix

Shananiganeer opened the pull request:
Add (boolean) "requireTool" data value to ore samples to control whether they need tools for their drops.
OreParts created with variants remove the non-variant MaterialPart from MaterialSystem after creating all variants.

Shananiganeer committed:
Add getBool

Added getBool method to facilitate the requireTool value for SubBlockOreSamplePart.

Shananiganeer committed:
Add requireTool PartDataPiece

Added (boolean) requireTool to control whether ore samples require appropriate tools.

Shananiganeer committed:
Add requireTool DataValue

(boolean) requireTool: if true, sets the return value of getMaterial() to Material.STONE and prevents onBlockActivated() from dropping items

Shananiganeer committed:
fix the oopsies

typos and prevent 'falling' samples from spawning drops if requireTool = true.

Shananiganeer committed:
OrePart bugfix

OreParts that were generated with variants would leave an orphaned non-variant MaterialPart that did not have an associated SubBlock and would cause a runtime error if getItemStack() was called.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/The-Acronym-Coders/BASE?pullRequest=99) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/The-Acronym-Coders/BASE?pullRequest=99) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/The-Acronym-Coders/BASE?pullRequest=99) it.</sub>

Shananiganeer committed:
Fix single line {}s

Shananiganeer committed:
Merge branch 'develop/1.12.0' of https://github.com/The-Acronym-Coders/BASE into develop/1.12.0

SkySom commented on src/main/java/com/teamacronymcoders/base/materialsystem/MaterialSystem.java:
```diff
@@ -105,6 +105,18 @@ public static void registerMaterialPart(MaterialPart materialPart) {
+    }
+    
+    public static void removeMaterialPart(MaterialPart materialPart) {
+        if (materialPartMap.containsKey(materialPart.getUnlocalizedName())) {
```
I don't believe you need to check it has something before you try and remove it, do you?

SkySom commented on src/main/java/com/teamacronymcoders/base/materialsystem/blocks/SubBlockOreSamplePart.java:
```diff
@@ -185,12 +194,12 @@ public boolean canPlaceBlockAt(World world, @Nonnull BlockPos pos) {
-        } else {
+            world.setBlockToAir(pos);
+        } else if (!requireTool) {
             spawnItemStackEntity(world, this.getItemStack().copy(), pos);
```
While poking around in here, is there a chance you could also take a quick look at replacing spawnItemStackEntity, with the method mentioned here: https://github.com/The-Acronym-Coders/BASE/issues/93

Shananiganeer committed:
PR changes

Removed unnecessary key verification from removeMaterialPart().
Changed SubBlockOreSamplePart.onNeighborChange() to use dropBlockAsItem() per #93.
